lib/Dialect/Arc/Transforms/LowerStateRewrite.cpp

```diff
@@ -0,0 +1,1206 @@
+  // region into it. Replace `llhd.halt` ops with `scf.yield`.
+  auto executeOp = module.finalBuilder.create<scf::ExecuteRegionOp>(
+      op.getLoc(), TypeRange{});
+  module.finalBuilder.cloneRegionBefore(op.getBody(), executeOp.getRegion(),
+                                        executeOp.getRegion().begin(), mapping);
+  executeOp.walk([&](llhd::HaltOp op) {
+    OpBuilder(op).create<scf::YieldOp>(op.getLoc());
+    op.erase();
+  });
+
+  return success();
+}
+
+/// Create the operations necessary to detect a posedge on the given clock,
+/// potentially reusing a previous posedge detection, and create an `scf.if`
+/// operation for that posedge. This also tries to reuse an `scf.if` operation
+/// immediately before the builder's insertion point if possible.
+scf::IfOp OpLowering::createIfClockOp(Value clock) {
+  auto &posedge = module.loweredPosedges[clock];
+  if (!posedge) {
+    auto loweredClock = lowerValue(clock, Phase::New);
+    if (!loweredClock)
+      return {};
+    posedge = module.detectPosedge(loweredClock);
+  }
+  return createOrReuseIf(module.builder, posedge, false);
+}
+
+//===----------------------------------------------------------------------===//
+// Value Lowering
+//===----------------------------------------------------------------------===//
+
+/// Lower a value being used by the current operation. This will mark the
+/// defining operation as to be lowered first (through `addPending`) in most
+/// cases. Some operations and values have special handling though. For example,
+/// states and memory reads are immediately materialized as a new read op.
+Value OpLowering::lowerValue(Value value, Phase phase) {
+  // Handle module inputs. They read the same in all phases.
+  if (auto arg = dyn_cast<BlockArgument>(value)) {
+    if (initial)
+      return {};
+    auto state = module.allocatedInputs[arg.getArgNumber()];
+    return module.getBuilder(phase).create<StateReadOp>(arg.getLoc(), state);
+  }
+
+  // Check if the value has already been lowered.
+  if (auto lowered = module.loweredValues.lookup({value, phase}))
+    return lowered;
+
+  // At this point the value is the result of an op. (Block arguments are
+  // handled above.)
+  auto result = cast<OpResult>(value);
+  auto *op = result.getOwner();
+
+  // Special handling for some ops.
+  if (auto instOp = dyn_cast<InstanceOp>(op))
+    return lowerValue(instOp, result, phase);
+  if (auto stateOp = dyn_cast<StateOp>(op))
+    return lowerValue(stateOp, result, phase);
+  if (auto dpiOp = dyn_cast<sim::DPICallOp>(op); dpiOp && dpiOp.getClock())
+    return lowerValue(dpiOp, result, phase);
+  if (auto readOp = dyn_cast<MemoryReadPortOp>(op))
+    return lowerValue(readOp, result, phase);
+  if (auto initialOp = dyn_cast<seq::InitialOp>(op))
+    return lowerValue(initialOp, result, phase);
+  if (auto castOp = dyn_cast<seq::FromImmutableOp>(op))
+    return lowerValue(castOp, result, phase);
```

Comment:
Nit: maybe use a TypeSwitch?

Reply:
Hmmm I'm not sure it would make this more readable. It generates the exact same code in the background (unfortunately it doesn't optimize anything really). The early returns are annoying to do in type switches :cry:.